crates/aptos/src/genesis/config.rs

```diff
@@ -55,13 +68,68 @@ pub struct ValidatorConfiguration {
     pub full_node_host: Option<HostAndPort>,
 }
 
+impl From<ValidatorConfiguration> for Validator {
+    fn from(config: ValidatorConfiguration) -> Self {
+        let auth_key = AuthenticationKey::ed25519(&config.account_key);
+        let validator_addresses = vec![config
+            .validator_host
+            .as_network_address(config.network_key)
+            .unwrap()];
+        let full_node_addresses = if let Some(full_node_host) = config.full_node_host {
+            vec![full_node_host
+                .as_network_address(config.network_key)
+                .unwrap()]
+        } else {
+            vec![]
+        };
+
+        Validator {
+            address: auth_key.derived_address(),
+            name: vec![], // TODO: To remove
+            consensus_pubkey: bcs::to_bytes(&config.consensus_key).unwrap(),
+            operator_address: auth_key.derived_address(),
+            operator_name: vec![], // TODO: To remove
+            network_address: bcs::to_bytes(&validator_addresses).unwrap(),
+            full_node_network_address: bcs::to_bytes(&full_node_addresses).unwrap(),
+            operator_auth_key: auth_key,
+            auth_key,
+        }
+    }
+}
+
 /// Combined Host (DnsName or IP) and port
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct HostAndPort {
     pub host: DnsName,
     pub port: u16,
 }
 
+impl HostAndPort {
+    pub fn as_network_address(&self, key: x25519::PublicKey) -> CliTypedResult<NetworkAddress> {
+        let host = self.host.to_string();
+
+        // Since DnsName supports IPs as well, let's properly fix what the type is
+        let host_protocol = if let Ok(ip) = Ipv4Addr::from_str(&host) {
+            Protocol::Ip4(ip)
+        } else if let Ok(ip) = Ipv6Addr::from_str(&host) {
+            Protocol::Ip6(ip)
+        } else {
+            Protocol::Dns(self.host.clone())
+        };
```

Comment:
I thought we had this code at some point that could take host:port and produce an appropriate string ... I guess not.

Reply:
We... might, but not one that converts it into the pieces used by NetworkAddress, which is a mess in itself.